fix: Filter invalid values in getIds

duncanbeevers opened the pull request:
## 🏋️ `getIds` robustness

`getIds` is a useful utility function, but it accepts a limited subset of the valid ObjectId constructor argument types, and throws when presented with invalid input.

It's not straight-forward to filter inputs to be valid constructor arguments; in some codebases we maintain bespoke utilities for performing these checks (eg; `isValidObjectId`).

Unfortunately, the canonical validation for whether one of these arguments is a valid argument to the `ObjectId` constructor is [cooked-in to the constructor itself](https://github.com/mongodb/js-bson/blob/26549d94c50c550d5fcdd08ec12da000b81a5074/src/objectid.ts#L87-L127).

Therefore, rather than re-implementing this logic (or trying to get it extracted and exposed upstream), I updated the `getIds` implementation to rely on the constructor logic through `try/catch`, and to discard invalid elements.

### 🐤 Backwards-Compatibility

<details>
<summary>Per <a href="https://github.com/plexinc/papr/pull/979#issuecomment-2541006143">discussion</a>, this is going to be a breaking change.</summary>

~In order to keep the change backwards-compatible, I gated the filtering functionality behind a `filterInvalid` option, passed to `getIds`. If we were okay with making a breaking release, I think this would be much nicer as the default behavior.~
</details>

### 🔧  Changes

I updated `getAll` (🟦) and its specs (🟡) in a few ways.

* [x] 🟦 make `getAll` accept an `Iterable<ObjectIdConstructorParameter>`, widening acceptable input types
* [x] 🟦 ~add optional `GetIdsOptions` argument to `getIds`, with optional `filterInvalid` boolean member~
* [x] 🟦 change `getIds` inner `map` to `flatMap`, to allow single-pass filtering
* [x] 🟦 add `try/catch` to `getIds` inner `flatMap`, to rely on ObjectId constructor validation behavior
* [x] 🟦 ~add error re-throw to `getIds` inner `flatMap`, gated on `filterInvalid` option~
* [x] 🟡 make expected result part of each test case
* [x] 🟡 introduce variety between test case values / fixtures
* [x] 🟡 add invariant check that expected result length must be less-than-or-equal-to input length
* [x] 🟡 add test cases for Uint8Array arguments
* [x] 🟡 add test cases for invalid input
* [x] 🟡 ~add test case for invalid input with no `filterInvalid` option~

duncanbeevers committed:
feat: Filter invalid ObjectId constructor types in getIds

ejmartin504 commented:
> If we were okay with making a breaking release, I think this would be much nicer as the default behavior.

I would much prefer this.

I know I'm always the one who says "is this _really_ a breaking change" -- and I must say this is an interesting example. After all, this should _fix_ existing uses, rather than break them... right?

avaly commented:
I agree with making this a breaking change. Let's remove the options argument.

duncanbeevers commented:
Is it worth it to extract the validation functionality from `bson` to avoid the `try/catch`, or ship with this implementation?

avaly reviewed (approved):
We can ship this as is, but let's mark the commit as a breaking change.

duncanbeevers commented:
I removed the `filterInvalid` option and made the filtering the default behavior. Let me know if there's more I need to do in order to mark this as a breaking change.

avaly commented on src/__tests__/utils.test.ts:
```diff
@@ -379,19 +386,90 @@ describe('utils', () => {
+      const [caseName, { input, expected }] = testCase;
+
+      test(`case ${caseName}`, (t: TestContext) => {
+        t.plan(4);
+        // Given
```
Since we're not doing anything async in this test, let's skip this plan.

avaly commented on src/__tests__/utils.test.ts:
```diff
@@ -379,19 +386,90 @@ describe('utils', () => {
+        t.plan(4);
+        // Given
+
+        t.assert.ok(expected.length <= input.length);
```
How about using the imports from `node:assert/strict` that we already do in this file instead of using `t.assert.`?

duncanbeevers replied:
We can do that, but I think we'd actually get more bang out of using the `TestContext` more generally.

That way you don't have to pick and choose which assertions to import; instead every test file and rely on the same basic import:

```ts
import { describe, test, TestContext } from 'node:test';
```

This not only reduces import boilerplate, but also means that if a test needs to grow async behavior we can add a `.plan` call without having to modify the rest of the case.

avaly reviewed (approved):
I think this PR should be fine as is. I don't think we need to mark it as breaking change. The only thing to change though is the commit prefix from `feat:` to `fix:` since I think this can be viewed as a bug fix.

nunomarks reviewed (approved):
CLGTM! 🎉

nunomarks committed:
fixup! improvements

nunomarks commented:
@duncanbeevers can we push this over the finish line? It's a nice addition.

I've addressed some feedback from Copilot/Claude and added some improvements, see if that looks good for you.

duncanbeevers commented:
> I've addressed some feedback from Copilot/Claude and added some improvements, see if that looks good for you.

LGTM. I can't merge until we get a different approval, but the code all looks wonderful.

<img width="849" height="130" alt="image" src="https://github.com/user-attachments/assets/436d0164-fd4f-417d-8755-9623fd6d015d" />

randhawk reviewed (approved):
Nice!
